spark_log_parser/extractor.py

```diff
@@ -0,0 +1,213 @@
+import os
+import boto3 as boto
+import gzip
+from pathlib import Path
+import shutil
+import struct
+import tarfile
+import zipfile
+
+import requests
+
+
+THRESHOLD_ENTRIES = 100
+THRESHOLD_SIZE = 5000000000
+THRESHOLD_RATIO = 100
+
+
+class Extractor:
+    """
+    Expands archive to a "work" directory and provides a list of extracted files
+    """
+
+    def __init__(self, source_url, work_dir):
+        self.source_url = source_url
+        self.work_dir = work_dir
+        self.file_total = 0
+        self.size_total = 0
+
+    def extract(self) -> list[Path]:
+        if self.source_url.scheme in {"https", "s3"}:
+            local_path = self.work_dir.joinpath(self.source_url.path.split("/")[-1])
```

Comment:
Alternative to method to get last field without hardcoding "/" character here: https://stackoverflow.com/questions/3925096/how-to-get-only-the-last-part-of-a-path-in-python

Reply:
`os.path` would work because we'll always be running on a 'nix server. But I don't like to use it on URL paths because it's intended for filesystem paths.